plugins/com.google.cloud.tools.eclipse.appengine.localserver/src/com/google/cloud/tools/eclipse/appengine/localserver/ui/GcpLocalRunTab.java

```diff
@@ -0,0 +1,290 @@
+        .applyTo(projectSelector);
+    GridLayoutFactory.fillDefaults().spacing(0, 0).generateLayout(projectSelectorComposite);
+
+    setControl(composite);
+
+    gcpIcon = SharedImages.GCP_IMAGE_DESCRIPTOR.createImage();
+    composite.addDisposeListener(new ImageDisposer(gcpIcon));
+  }
+
+  private void updateProjectSelector() {
+    Credential credential = accountSelector.getSelectedCredential();
+    if (credential == null) {
+      projectSelector.setProjects(new ArrayList<GcpProject>());
+    } else {
+      try {
+        List<GcpProject> gcpProjects = projectRepository.getProjects(credential);
+        projectSelector.setProjects(gcpProjects);
+      } catch (ProjectRepositoryException e) {
+        logger.log(Level.WARNING,
+            "Could not retrieve GCP project information from server.", e); //$NON-NLS-1$
+      }
+    }
+  }
+
+  @Override
+  public void setDefaults(ILaunchConfigurationWorkingCopy configuration) {
+    // No particular default values to set in a newly created configuration.
+  }
+
+  @Override
+  public void activated(ILaunchConfigurationWorkingCopy workingCopy) {
+    activated = true;
+    super.activated(workingCopy);
+  }
+
+  @Override
+  public void deactivated(ILaunchConfigurationWorkingCopy workingCopy) {
+    super.deactivated(workingCopy);
+    activated = false;
+  }
+
+  @Override
+  public void initializeFrom(ILaunchConfiguration configuration) {
+    accountEmailModel = getAttribute(configuration, ATTRIBUTE_ACCOUNT_EMAIL, ""); //$NON-NLS-1$
+
+    Map<String, String> environmentMap = getEnvironmentMap(configuration);
+    gcpProjectIdModel = Strings.nullToEmpty(environmentMap.get(PROJECT_ID_ENVIRONMENT_VARIABLE));
+    String serviceKey = Strings.nullToEmpty(environmentMap.get(SERVICE_KEY_ENVIRONMENT_VARIABLE));
+
+    initializingUiValues = true;
+    accountSelector.selectAccount(accountEmailModel);
+    projectSelector.selectProjectId(gcpProjectIdModel);
+    serviceKeyInput.setText(serviceKey);
+    initializingUiValues = false;
+  }
+
+  @Override
+  public void performApply(ILaunchConfigurationWorkingCopy configuration) {
+    if (activated) {
+      configuration.setAttribute(ATTRIBUTE_ACCOUNT_EMAIL, accountEmailModel);
+
+      Map<String, String> environmentMap = getEnvironmentMap(configuration);
+      if (!gcpProjectIdModel.isEmpty()) {
+        environmentMap.put(PROJECT_ID_ENVIRONMENT_VARIABLE, gcpProjectIdModel);
+      } else {
+        environmentMap.remove(PROJECT_ID_ENVIRONMENT_VARIABLE);
+      }
+      if (!serviceKeyInput.getText().isEmpty()) {
+        environmentMap.put(SERVICE_KEY_ENVIRONMENT_VARIABLE, serviceKeyInput.getText());
+      } else {
+        environmentMap.remove(SERVICE_KEY_ENVIRONMENT_VARIABLE);
+      }
+      configuration.setAttribute(ILaunchManager.ATTR_ENVIRONMENT_VARIABLES, environmentMap);
+
+      if (environmentTab != null) {
+        // Needed to make the "Environment" tab to re-initialize its UI from the changes made here.
```

Comment:
It's not clear if it is a bug. Its comment says it won't do anything when activated/deactivated, which is OK from their perspective because everything just works when isolated. It's not working because we are modifying their model in a different tab.

Reply:
But it's not _their_ model: it's an `ILaunchConfiguration` that is shared across the tabs.

Reply:
Filed: https://bugs.eclipse.org/bugs/show_bug.cgi?id=527142